langserver/handler.go

```diff
@@ -71,19 +67,24 @@ func (h *LangHandler) resetCaches(lock bool) {
 	if lock {
 		h.mu.Lock()
 	}
-	h.cache = map[typecheckKey]*typecheckResult{}
+
 	h.importGraphOnce = sync.Once{}
 	h.importGraph = nil
-	if lock {
-		h.mu.Unlock()
+
+	if h.typecheckCache == nil {
+		h.typecheckCache = newTypecheckCache()
+	} else {
+		h.typecheckCache.Purge()
```

Comment:
yes it is weird!! this whole bit of the code is one of the things I was aiming to cleanup this week. Basically we rely on resetCaches being called when initialize request is sent. Before initialize is sent everything is nil and we enforce initialize is the first request. Before initialize is done, every other request is either blocked or rejected. Then this bit of code is used again for clearing caches when the user edits a file (not used in production, but used in desktop environments).

Its all pretty crappy TBH. The major change I wanted to make in arch was separating out the init from the reset, so we don't have think about protecting assignment of every variable in LangHandler.